# 1.2 Exercise: Explore a Pandas Dataframe
# Rahul Rajeev

In [87]:
# import
import pandas as pd
from collections import Counter
import numpy as np

In [13]:
# 1. Load data set into pandas
vgsales = pd.read_csv('videogamessales2016.csv', delimiter=',')

In [15]:
# 2. Display the first 10 rows
vgsales.head(10)

Name Platform  Year_of_Release         Genre  \
0                 Wii Sports      Wii           2006.0        Sports   
1          Super Mario Bros.      NES           1985.0      Platform   
2             Mario Kart Wii      Wii           2008.0        Racing   
3          Wii Sports Resort      Wii           2009.0        Sports   
4   Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
5                     Tetris       GB           1989.0        Puzzle   
6      New Super Mario Bros.       DS           2006.0      Platform   
7                   Wii Play      Wii           2006.0          Misc   
8  New Super Mario Bros. Wii      Wii           2009.0      Platform   
9                  Duck Hunt      NES           1984.0       Shooter   

  Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  \
0  Nintendo     41.36     28.96      3.77         8.45         82.53   
1  Nintendo     29.08      3.58      6.81         0.77         40.24   
2  Nintendo     15.68     12.76      3.79         3.29         35.52   
3  Nintendo     15.61     10.93      3.28         2.95         32.77   
4  Nintendo     11.27      8.89     10.22         1.00         31.37   
5  Nintendo     23.20      2.26      4.22         0.58         30.26   
6  Nintendo     11.28      9.14      6.50         2.88         29.80   
7  Nintendo     13.96      9.18      2.93         2.84         28.92   
8  Nintendo     14.44      6.94      4.70         2.24         28.32   
9  Nintendo     26.93      0.63      0.28         0.47         28.31   

   Critic_Score  Critic_Count User_Score  User_Count Developer Rating  
0          76.0          51.0          8       322.0  Nintendo      E  
1           NaN           NaN        NaN         NaN       NaN    NaN  
2          82.0          73.0        8.3       709.0  Nintendo      E  
3          80.0          73.0          8       192.0  Nintendo      E  
4           NaN           NaN        NaN         NaN       NaN    NaN  
5           NaN           NaN        NaN         NaN       NaN    NaN  
6          89.0          65.0        8.5       431.0  Nintendo      E  
7          58.0          41.0        6.6       129.0  Nintendo      E  
8          87.0          80.0        8.4       594.0  Nintendo      E  
9           NaN           NaN        NaN         NaN       NaN    NaN

In [19]:
# 3. dimensions of the dataset, and explain the context

# use shape to get out the rows and columns tuple
nrow, ncol = vgsales.shape
print(nrow, ncol)

16719 16


**Thoughts:** In the context of this dataset, there are 16 different attributes for video games, and there are a total of 16719 games.

In [24]:
# 4. Top 5 games by critic score

# sort_values with ascending parameter to be false
topcritic = vgsales.sort_values(by=['Critic_Score'], ascending = False)
topcritic.head()

Name Platform  Year_of_Release     Genre  \
227   Tony Hawk's Pro Skater 2       PS           2000.0    Sports   
57         Grand Theft Auto IV      PS3           2008.0    Action   
51         Grand Theft Auto IV     X360           2008.0    Action   
5350               SoulCalibur       DC           1999.0  Fighting   
165         Grand Theft Auto V     XOne           2014.0    Action   

                 Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  \
227             Activision      3.05      1.41      0.02         0.20   
57    Take-Two Interactive      4.76      3.69      0.44         1.61   
51    Take-Two Interactive      6.76      3.07      0.14         1.03   
5350    Namco Bandai Games      0.00      0.00      0.34         0.00   
165   Take-Two Interactive      2.81      2.19      0.00         0.47   

      Global_Sales  Critic_Score  Critic_Count User_Score  User_Count  \
227           4.68          98.0          19.0        7.7       299.0   
57           10.50          98.0          64.0        7.5      2833.0   
51           11.01          98.0          86.0        7.9      2951.0   
5350          0.34          98.0          24.0        8.8       200.0   
165           5.48          97.0          14.0        7.9       764.0   

                    Developer Rating  
227   Neversoft Entertainment      T  
57             Rockstar North      M  
51             Rockstar North      M  
5350                    Namco      T  
165            Rockstar North      M

In [37]:
# 5. Find the number of video games in each genre

# create a list of genre entries from dataframe
genre = list(vgsales['Genre'])

# use counter to display the genres and counts as a dictionary
genrecounts = Counter(genre)

# revert to dataframe for easy viewing
genredf = pd.DataFrame.from_dict(genrecounts, orient='index')
genredf

0
Sports        2348
Platform       888
Racing        1249
Role-Playing  1500
Puzzle         580
Misc          1750
Shooter       1323
Simulation     874
Action        3370
Fighting       849
Adventure     1303
Strategy       683
NaN              2

In [42]:
# 6. Find the first five games in the dataframe on the SNES platform

# creating subset
snesvg = vgsales[vgsales['Platform'] == 'SNES']
snesvg.head()

# sort by with ascending = True, then finding the first 5
sortedsnes = snesvg.sort_values(by='Year_of_Release')
sortedsnes.head()

Name Platform  Year_of_Release       Genre Publisher  \
18    Super Mario World     SNES           1990.0    Platform  Nintendo   
1195        Final Fight     SNES           1990.0      Action    Capcom   
511              F-Zero     SNES           1990.0      Racing  Nintendo   
1791         Pilotwings     SNES           1990.0  Simulation  Nintendo   
1111      Super Scope 6     SNES           1991.0     Shooter  Nintendo   

      NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  Critic_Score  \
18       12.78      3.75      3.54         0.55         20.61           NaN   
1195      0.67      0.17      0.69         0.03          1.56           NaN   
511       1.37      0.51      0.89         0.07          2.85           NaN   
1791      0.46      0.17      0.48         0.02          1.14           NaN   
1111      1.06      0.38      0.15         0.05          1.65           NaN   

      Critic_Count User_Score  User_Count Developer Rating  
18             NaN        NaN         NaN       NaN    NaN  
1195           NaN        NaN         NaN       NaN    NaN  
511            NaN        NaN         NaN       NaN    NaN  
1791           NaN        NaN         NaN       NaN    NaN  
1111           NaN        NaN         NaN       NaN    NaN

In [70]:
# 7. Find the five publishers with the highest total global sales. 
# Note: You will need to calculate the total global sales for each publisher to do this.

# group by publisher, sum the global sales column and reset the index to keep as dataframe
global_sales_by_pub = vgsales.groupby(by='Publisher')['Global_Sales'].sum().reset_index()

# sorting the values by global sales, ascending is false, and displaying first 5
global_sales_by_pub = global_sales_by_pub.sort_values(by = 'Global_Sales', ascending=False).head(5)
global_sales_by_pub

Publisher  Global_Sales
361                     Nintendo       1788.81
139              Electronic Arts       1116.96
21                    Activision        731.16
458  Sony Computer Entertainment        606.48
528                      Ubisoft        471.61

In [78]:
# 8. Create a new column in the data frame that calculates the percentage of global sales from North America. 
# Display the first five rows of the new data frame.

# performing the operation and equating it to a new column in the dataframe
vgsales['NA_Percent'] = vgsales['NA_Sales'] / vgsales['Global_Sales']
vgsales.head()

Name Platform  Year_of_Release         Genre Publisher  \
0                Wii Sports      Wii           2006.0        Sports  Nintendo   
1         Super Mario Bros.      NES           1985.0      Platform  Nintendo   
2            Mario Kart Wii      Wii           2008.0        Racing  Nintendo   
3         Wii Sports Resort      Wii           2009.0        Sports  Nintendo   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  Critic_Score  \
0     41.36     28.96      3.77         8.45         82.53          76.0   
1     29.08      3.58      6.81         0.77         40.24           NaN   
2     15.68     12.76      3.79         3.29         35.52          82.0   
3     15.61     10.93      3.28         2.95         32.77          80.0   
4     11.27      8.89     10.22         1.00         31.37           NaN   

   Critic_Count User_Score  User_Count Developer Rating  NA_Percent  
0          51.0          8       322.0  Nintendo      E    0.501151  
1           NaN        NaN         NaN       NaN    NaN    0.722664  
2          73.0        8.3       709.0  Nintendo      E    0.441441  
3          73.0          8       192.0  Nintendo      E    0.476350  
4           NaN        NaN         NaN       NaN    NaN    0.359260

In [81]:
# 9. Find the number NaN entries (missing data values) in each column.
vgsales.isnull().sum().reset_index()

index     0
0              Name     2
1          Platform     0
2   Year_of_Release   269
3             Genre     2
4         Publisher    54
5          NA_Sales     0
6          EU_Sales     0
7          JP_Sales     0
8       Other_Sales     0
9      Global_Sales     0
10     Critic_Score  8582
11     Critic_Count  8582
12       User_Score  6704
13       User_Count  9129
14        Developer  6623
15           Rating  6769
16       NA_Percent     0

In [97]:
# 10. Try to calculate the median user score of all the video games. 
# You will likely run into an error because some of the user score entries are 
# a non-numerical string that cannot be converted to a float. 
# Find and replace this string with NaN and then calculate the median. 
# Then, replace all NaN entries in the user score column with the median value.

# setting column of user scores to a variable
user_score = vgsales['User_Score']

# cleaning everything but a number using regex
user_score_clean = user_score.replace(r'^([A-Za-z]|[0-9]|_)+$', np.NaN, regex=True)

# demonstrating that the NaN count has increased from strings being replaced
print('Before:', user_score.isnull().sum(), 'After:',user_score_clean.isnull().sum())

Before: 6704 After: 10012


In [98]:
# replacing all the NaN values with the median of the cleaned column
user_score_clean = user_score.replace(np.NaN, user_score_clean.median())

# double checking the count after
print(user_score_clean.isnull().sum())

0


In [100]:
# replacing the existing user score column with the cleaned version
vgsales['User_Score'] = user_score_clean
vgsales.head()

Name Platform  Year_of_Release         Genre Publisher  \
0                Wii Sports      Wii           2006.0        Sports  Nintendo   
1         Super Mario Bros.      NES           1985.0      Platform  Nintendo   
2            Mario Kart Wii      Wii           2008.0        Racing  Nintendo   
3         Wii Sports Resort      Wii           2009.0        Sports  Nintendo   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  Critic_Score  \
0     41.36     28.96      3.77         8.45         82.53          76.0   
1     29.08      3.58      6.81         0.77         40.24           NaN   
2     15.68     12.76      3.79         3.29         35.52          82.0   
3     15.61     10.93      3.28         2.95         32.77          80.0   
4     11.27      8.89     10.22         1.00         31.37           NaN   

   Critic_Count User_Score  User_Count Developer Rating  NA_Percent  
0          51.0          8       322.0  Nintendo      E    0.501151  
1           NaN        7.5         NaN       NaN    NaN    0.722664  
2          73.0        8.3       709.0  Nintendo      E    0.441441  
3          73.0          8       192.0  Nintendo      E    0.476350  
4           NaN        7.5         NaN       NaN    NaN    0.359260

**Thoughts:** Here, to prove that the code did indeed work, I double checked with an earlier call to head() on the dataframe before the strings and NaN were replaced with median, and found that the Super Mario Bros on the NES now has a user_score of 7.5, the median of the user_scores. 